### ISSUE TITLE
Question about Training Loss

### ISSUE THREAD

**Issue by RMSnow** [mct10/RepCodec#6]
Thanks for your excellent work!

I'm training a RepCodec based on HuBERT-large (layer18) 1024-dim features from scratch. However, the training loss tend to be NAN:

```
2024-08-19 20:53:50 | INFO | repcodec_train | (Steps: 2100) train/train/ppl = 0.0000.                                                            
2024-08-19 20:53:50 | INFO | repcodec_train | (Steps: 2100) train/train/vqloss = 0.0000.                                                         
2024-08-19 20:53:50 | INFO | repcodec_train | (Steps: 2100) train/reconstruct_loss = nan.                                                        
2024-08-19 20:53:50 | INFO | repcodec_train | (Steps: 2100) train/codec_loss = 0.0000.                                                           
2024-08-19 20:53:50 | WARNING | tensorboardX.x2num | NaN or Inf found in input tensor.                                                           
[train]:   1%|█                                                                                          | 2199/200000 [02:13<2:25:43, 22.62it/s]
2024-08-19 20:53:55 | INFO | repcodec_train | (Steps: 2200) train/train/ppl = 0.0000.                                                            
2024-08-19 20:53:55 | INFO | repcodec_train | (Steps: 2200) train/train/vqloss = 0.0000.                                                         
2024-08-19 20:53:55 | INFO | repcodec_train | (Steps: 2200) train/reconstruct_loss = nan.
2024-08-19 20:53:55 | INFO | repcodec_train | (Steps: 2200) train/codec_loss = 0.0000.
2024-08-19 20:53:55 | WARNING | tensorboardX.x2num | NaN or Inf found in input tensor.
```

Here is my model setting (I think it is same with https://github.com/mct10/RepCodec/blob/main/repcodec/configs/repcodec_dim1024.yaml):
```
model_params:
    input_channels: 1024
    output_channels: 1024
    encode_channels: 1024
    decode_channels: 1024
    code_dim: 1024
    codebook_num: 1
    codebook_size: 1024
    bias: true
    enc_ratios: [1, 1]
    dec_ratios: [1, 1]
    enc_strides: [1, 1]  # no downsampling
    dec_strides: [1, 1]
    enc_kernel_size: 3
    dec_kernel_size: 3
    enc_block_dilations: [1, 1]
    enc_block_kernel_size: 3
    dec_block_dilations: [1, 1]
    dec_block_kernel_size: 3
```

FYI, I use the same training data as your pretrained model (Librispeech train-clean-100).

**Comment by RMSnow** [comment]
By the way, when I want to increase the codebook_size to 2048, the training curves are:

<img width="1546" alt="image" src="https://github.com/user-attachments/assets/3df2ec82-02a1-4ddb-9d20-d9bc161d3206">

I found the reconstruct loss was cycling, and the codec_loss, ppl, and vq_loss are always 0. Moreover, when evaluating, I also got the NAN for reconstruct loss:

<img width="1542" alt="image" src="https://github.com/user-attachments/assets/ddfd89da-0cc3-47f2-80d5-a1e8ec62faf4">

**Comment by mct10** [comment]
Hi.
I tested the training myself just now, and it worked fine. I also had a look at your scripts and they seem also correct. So I am afraid I am not sure what went wrong in your case, for now...
I am wondering, for the case when codebook_size is 1024, is the training loss 0/NAN from step 0? Or the training loss becomes NAN suddenly at step 2100?

**Comment by RMSnow** [comment]
Hi @mct10, thanks for your quick reply! I figured out the bug. This is caused by my HuBERT-large features extraction. Now I have trained the RepCodec at different codebook_size (1024, 2048, 4096, ... 16384), and there are the training losses:

<img width="1471" alt="image" src="https://github.com/user-attachments/assets/5dbcf6e0-c804-4258-a3ec-6840452fa89b">

I have some questions to consult:

1. Are these training losses normal and similar to yours?

2. I found the scale of reconstruct loss and vqloss are so large (e.g.: the converged reconstruct loss is about 900 / 45.0 = 20, vqloss is around 170). Is this normal, since in my previous experience, the L2 loss might be very small? 
I guess the reasons are related to the value range of HuBERT-large features (I checked it and found it owned some large float values like over 10). Therefore, I wonder to know did you try some normalization tricks for these features during preprocessing? Will it work?

3. I found that for different codebook_size, the reconstruct loss and vqloss are almost same. The only difference is PPL. The larger codebook_size will have a higher PPL. Is this conclusion same to your experiments? Do you have any insights about that?

Thanks again, for your great work and high-quality code!

**Comment by mct10** [comment]
I am glad you solved the issue!
For your questions:
1. Yes, the losses look good to me.
2. Yes, I tried to apply normalization to Whisper medium and HuBERT base features by dividing all features with their averaged L2 norm. And yes, after applying normalization, you can get a much smaller reconstruction loss (close to 0). But the ASR performance was basically the same, so we didn't apply it in the end.
3. It seems reasonable to me that a larger codebook_size will have a higher PPL. 
- My understanding is, the PPL here is a measurement of how likely the indicies appear. The code is
`perplexity = torch.exp(-torch.sum(avg_probs * torch.log(avg_probs + 1e-10)))`
The `avg_probs` here is a vector of size `(codebook_size, )` and each entry is the probability of that index, meaning `avg_probs[i] = p(index=i) = count(index=i)/count(all index).`
So, you know, the upperbound of `perplexity` is  `codebook_size` when all the indices appear equally likely. When `codebook_size` increases, the upperbound of `perplexity` also increases.
- Also, I think the model is in trouble when only a few indices appear. So, our goal is (maybe) to have a relatively high PPL. 
- In conclusion, I think your result makes sense. And the PPL _might_ tell us something about whether to increase/decrease the codebook size (although I guess we still need downstream tasks to confirm that). For example, c16384 has a lower PPL than c4096 even its upper bound is higher, which might indicate that 16384 is too large and not all of the entries have been used.

At last, just some random thoughts. Feel free to correct me if I'm wrong :)

**Comment by RMSnow** [comment]
Thank you very much! Appreciate again for this work!
